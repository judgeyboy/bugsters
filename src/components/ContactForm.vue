<template>
  <div>
    <form class="form-horizontal"
          @submit.prevent="handleSubmit('contact-form')"
          data-vv-scope="contact-form">

      <div class="form-group">
        <div class="col-3">
          <label class="form-label" for="name">Name</label>
        </div>
        <div class="col-9">
          <input type="text"
                 id="name"
                 class="form-input"
                 name="name"
                 placeholder="name"
                 v-model="name"
                 v-validate="'required'"
                 :class="{'is-error': errors.has('contact-form.name')}">
        </div>
      </div>

      <div class="form-group">
        <div class="col-3">
          <label class="form-label" for="email">Email</label>
        </div>
        <div class="col-9">
          <input type="email"
                 id="email"
                 class="form-input"
                 name="email"
                 placeholder="email"
                 v-model="email"
                 v-validate="'required|email'"
                 :class="{'is-error': errors.has('contact-form.email')}">
        </div>
      </div>

      <div class="form-group">
        <div class="col-3">
          <label class="form-label" for="subject">Subject</label>
        </div>
        <div class="col-9">
          <input type="text"
                 id="subject"
                 class="form-input"
                 name="subject"
                 placeholder="subject"
                 v-model="subject"
                 v-validate="'required'"
                 :class="{'is-error': errors.has('contact-form.subject')}">
        </div>
      </div>

      <div class="form-group">
        <div class="col-3">
          <label class="form-label" for="message">Message</label>
        </div>
        <div class="col-9">
          <textarea placeholder="message"
                    id="message"
                    class="form-input"
                    name="message"
                    v-model="message"
                    v-validate="'required|max:200'"
                    :class="{'is-error': errors.has('contact-form.message')}"></textarea>
        </div>
      </div>

      <input type="submit" value="Send" class="btn btn-primary">
    </form>
  </div>
</template>

<script>
export default {

  name: 'ContactForm',

  data () {
    return {
      name: '',
      email: '',
      subject: '',
      message: ''
    }
  },

  inject: ['$validator'],

  methods: {
    handleSubmit (scope) {
      return this.$validator.validateAll(scope).then(valid => {
        if (!valid) {
          alert('fix dem errors')
          return
        }

        const contactDetails = {
          name: this.name,
          email: this.email,
          subject: this.subject,
          message: this.message
        }

        this.$emit('submit', contactDetails)
      })
    }
  }

}
</script>
