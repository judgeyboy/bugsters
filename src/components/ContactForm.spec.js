import Vue from 'vue'
import VeeValidate, { Validator } from 'vee-validate'
import { shallow } from 'vue-test-utils'

import ContactForm from './ContactForm'

Vue.use(VeeValidate)

describe('ContactForm component', () => {

  it('should be a vue instance', () => {
    const wrapper = shallow(ContactForm, {
      provide: {
        $validator: new Validator()
      }
    })

    expect(wrapper.isVueInstance()).toBe(true)
  })

  it('should call handleSubmit when clicking the send button', () => {
    const wrapper = shallow(ContactForm, {
      provide: { $validator: new Validator() }
    })
    let spy = jest.spyOn(wrapper.vm, 'handleSubmit')

    const submitButton = wrapper.find('input[type=submit]')
    submitButton.trigger('click')

    expect(spy).toHaveBeenCalledWith('contact-form')
  })

  it('should emit "submit" event with contact details', async () => {
    const wrapper = shallow(ContactForm, {
      provide: { $validator: new Validator() }
    })
    let spy = jest.spyOn(wrapper.vm, '$emit')

    const contactDetails = {
      name: 'Sterling Archer',
      email: 'duchess@spy-agency.com',
      subject: 'Danger Zone',
      message: 'This is an awesome message'
    }

    wrapper.setData(contactDetails)

    await wrapper.vm.handleSubmit('contact-form')

    expect(spy).toHaveBeenCalledWith('submit', contactDetails)
  })

  it('should not emit "submit" event if the form has errors', async () => {
    const wrapper = shallow(ContactForm, {
      provide: { $validator: new Validator() }
    })
    let spy = jest.spyOn(wrapper.vm, '$emit')

    const invalidContactDetails = {
      name: 'Sterling Archer',
      email: 'duchess@spy-agency.com',
      subject: 'Danger Subject',
      message: ''
    }

    wrapper.setData(invalidContactDetails)

    await wrapper.vm.handleSubmit('contact-form')

    expect(spy).not.toHaveBeenCalled()
  })

})
