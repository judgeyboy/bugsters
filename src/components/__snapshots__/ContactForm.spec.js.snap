// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm component snapshot 1`] = `"<div class=\\"contact-form\\"><form><input type=\\"text\\" name=\\"name\\" placeholder=\\"name\\" class=\\"\\" data-vv-id=\\"_3dx17ydnq\\" aria-required=\\"true\\" aria-invalid=\\"false\\"> <br> <input type=\\"email\\" name=\\"email\\" placeholder=\\"email\\" class=\\"\\" data-vv-id=\\"_3b96kxnbv\\" aria-required=\\"true\\" aria-invalid=\\"false\\"> <br> <input type=\\"text\\" name=\\"subject\\" placeholder=\\"subject\\" class=\\"\\" data-vv-id=\\"_qlido71zb\\" aria-required=\\"true\\" aria-invalid=\\"false\\"> <br> <textarea placeholder=\\"message\\" name=\\"message\\" class=\\"\\" data-vv-id=\\"_s2lk782yz\\" aria-required=\\"true\\" aria-invalid=\\"false\\"></textarea> <br> <input type=\\"submit\\" value=\\"Send\\" class=\\"button\\"></form></div>"`;
